src/c/realize_c_type.c

```diff
@@ -778,7 +778,7 @@ static int do_realize_lazy_struct(CTypeDescrObject *ct)
 {
     /* This is called by force_lazy_struct() in _cffi_backend.c */
     assert(ct->ct_flags & (CT_STRUCT | CT_UNION));
-
+    assert(!(ct->ct_flags_mut & CT_UNDER_CONSTRUCTION));
```

Comment:
This seems like it might be a temporary assert. I think we can end up with concurrent calls to `do_realize_lazy_struct()` on the same `CTypeDescrObject`

Reply:
I think @kumaraditya303 plans to turn the `CT_UNDER_CONSTRUCTION` flag into a uint8 member of the `CTypeDescrObject` struct to facilitate reading and writing to it using atomic operations implemented using something like `npy_atomic.h`, which is a bare-bones version of `pyatomic.h`.

When that refactor is ready this assert can be moved or deleted when we're sure concurrent calls to `b_complete_struct_or_union` are safe.